02_bubble/ailyntang/sort.go

```diff
@@ -0,0 +1,52 @@
+package main
+
+import (
+	"fmt"
+	"io"
+	"os"
+)
+
+var out io.Writer = os.Stdout
+
+func main() {
+	fmt.Fprintln(out, bubble([]int{3, 2, 1, 5}))
+}
+
+func bubble(s []int) []int {
+	if len(s) < 2 {
+		return s
+	}
+
+	beforeSort := make([]int, len(s))
+	copy(beforeSort, s)
+	afterSort := make([]int, len(s))
+	copy(afterSort, s)
```

Comment:
It seems to me that you want to separate copying of the input (so you don't mutate the original input) and the sorting algorithm, which I think is a great idea.
You could do something like:

```go
func bubbleSort(s []int) []int {
	scopy := make([]int, len(s))
	copy(scopy, s)
	bubbleSortInPlace(scopy)
	return scopy
}

func bubbleSortInPlace(s []int) {
	for i := len(s); i > 0; i-- {
              // ...
	}
}

```

Reply:
Thanks for the feedback @juliaogris . You've got it right, I was trying to avoid mutating the input. I thought about how I could reduce the copies of the input from two slices to one, but based on the way I've implemented the sort, I don't think it will work unless I'm happy to mutate the original input.

I've added some comments to my code to make it more readable and to explain the difference between `bubble` and `sortAdjacentPairs`. 

It's different to the bubble sorting algorithms listed on the wiki. I look forward to hearing your thoughts. If it's still confusing, I'm happy to change it to a more traditional algorithm. 

Thanks again for reviewing.